scarlet/wavelet.py

```diff
@@ -28,91 +28,82 @@ def __init__(self, image, lvl = None, starlet = None, direct = False):
             if set to False, the transform is performed by convolving the image by the wavelet transform of a dirac.
         """
-        # Shape for the starlet padding. It is also an fft fast shape.
         self.seed = None
-        if starlet is None:
-            self._starlet_shape = get_starlet_shape(image.shape, lvl = lvl)
-        else:
-            self._starlet_shape = starlet.shape
-        # Image (as a cube)
-        if len(image.shape) == 2:
-            self._image = image[np.newaxis, :, :]
-        else:
-            self._image = image
-        # Original shape of the image
-        self._image_shape = self._image.shape
-        # Can be initialised by a starlet transform to perform inverse transforms
-        if starlet is None:
-            if direct == True:
-                self._starlet = self.direct_transform()
+        # Transform method
+        self._direct = direct
+        if coefficients is None:
+            if image is None:
+                raise InputError('At least an image or a set of coefficients should be provided')
             else:
-                self._starlet = self.transform()
+                if len(image.shape) == 2:
+                    self._image = image[np.newaxis, :, :]
+                else:
+                    self._image = image
+                # Original shape of the image
+                self._image_shape = self._image.shape
+                # Padding shape for the starlet transform
+                if lvl is None:
+                    self._lvl = get_starlet_shape(image.shape)
+                else:
+                    self._lvl = lvl
+            self._coefficients = None
+
         else:
-            if len(np.shape(starlet)) == 3:
-                self._starlet = starlet[np.newaxis, :, :, :]
+            self._image_shape = [coefficients.shape[0], *coefficients.shape[-2:]]
+            if len(np.shape(coefficients)) == 3:
+                self._coefficients = coefficients[np.newaxis, :, :, :]
+            else:
+                self._coefficients = coefficients
+            self._lvl = self._coefficients.shape[1]
+            if image is None:
+                self._image = None
             else:
-                self._starlet = starlet
+                if len(image.shape) == 2:
+                    self._image = image[np.newaxis, :, :]
+                else:
+                    self._image = image
 
+        self._starlet_shape = [self._lvl, *self._image_shape[-2:]]
         if self.seed is None:
             self.seed = mk_starlet(self._starlet_shape)
         self._norm = np.sqrt(np.sum(self.seed ** 2, axis=(-2, -1)))
 
     @property
     def image(self):
         """The real space image"""
-        return self._image
+        if self._image is None:
+            rec = []
+            for star in self._coefficients:
+                rec.append(iuwt(star))
+            return np.array(rec)
+        else:
+            return self._image
```

Comment:
I'd remove the if/else in favor of always computing the inverse transform. Unless there is a good reason the store and maintain it.

Reply:
So, as you've seen, my whole initial plan was to have a rather rigid structure for the class to prevent computing transforms and inverse when it's already been done. I surrender to the argument of memory. The coefficients and images are in this class only to be transformed and modify. There is therefore no reason to keep them as we do in the Fourier class where it is useful to keep the psf and a few other operators in memory. I'll make it always compute